Add tox for testing and Python 3.5 test support

bennylope opened the pull request:
Creates isolated test environments for conveniently testing against multiple Python versions.

The tox configuration here tests against Python 2.7, Python 3.4, Python 3.5, PyPy, and PyPy3 by default. Given that the issues referenced in #114 and #117 have to do with build isolation, this should resolve those issues (have not experienced them, that much I'm certain about).

kevinhughes27 commented:
Can you give a bit more context here? Is this so that I can test against multiple python versions locally?

How would I actually use this locally?

How does this specifically fix issues with people's local setup as in the issues you described? Is tox a common utility in the python community (sorry haven't spent much time working with python the last few years)

I am a bit against adding more tooling here since it adds complexity and we're not python experts so simple is better!

bennylope commented:
> Is this so that I can test against multiple python versions locally?

Yes, that's the *primary* purpose of using tox. The way it does this is by creating and managing virtual environments for *each* version of Python (or each environment, which may include a matrix of environments, much like Travis).

> How would I actually use this locally?

You'd run the `tox` command. I added some [documentation to the README](https://github.com/Shopify/shopify_python_api/pull/134/files#diff-04c6e90faac2675aa89e2176d2eec7d8) to show how to use this.

I just removed `pypy` and `pypy3` from the default environment set, since that could cause confusion. When you run the `tox` command it will test against each specified environment. In this case it will look for a Python 2.7 interpreter, a Python 3.4 interpreter, and a Python 3.5 interpreter, build an isolated virtualenv for each Python version, and run the tests in that isolated virtualenv.

> How does this specifically fix issues with people's local setup as in the issues you described?

I'm not 100% sure it will, but relatively confident. It's creating isolated [virtualenvs](http://docs.python-guide.org/en/latest/dev/virtualenvs/), separate from whatever other Python virtualenv they're working with. 

> Is tox a common utility in the python community?

Yes. The best part about adding tox support is that _no one is required to use it_. It makes it very easy to test against multiple versions of Python right away so you don't make incompatible changes, but it works by running whatever test script you feed it. Here it's running `python setup.py test`.

> I am a bit against adding more tooling here since it adds complexity and we're not python experts so simple is better!

This, then, is the type of tooling you _do_ want to add! No one wants to manually test against every supported Python version. It's simple to set up, there's no requirement to use it, and it lets you test against Python versions without having to think about it.

bennylope commented:
Looking through the original issue again and my own local codebase, `pbr` is required by `mock`. A secondary change would be to update the mock version required, however I would strongly recommend adding tox support for any Python project supporting multiple versions of Python.

What other projects are using tox to make testing easier?

- [Flask](https://github.com/pallets/flask/blob/master/tox.ini)
- [Jinja2](https://github.com/pallets/jinja/blob/master/tox.ini)
- [Ansible](https://github.com/ansible/ansible/blob/devel/tox.ini)
- [speedtest-cli](https://github.com/sivel/speedtest-cli/blob/master/tox.ini)
- [Rackspace Cloud Files API](https://github.com/rackspace/pyrax/blob/master/tox.ini)
- [SQLAlchemy](https://bitbucket.org/zzzeek/sqlalchemy/src/35e93db7ffb5806df22e997fde1f966fa8c453b0/tox.ini?fileviewer=file-view-default)

And soon... shopify_python_api! :dancers:

kevinhughes27 commented:
Thanks for the detailed explanation. Seems like it would be useful so I am okay with this! Thanks for putting this together and explaining it.
